src/mapillary.py

```diff
@@ -0,0 +1,145 @@
+            return {
+                "latitude": latitude,
+                "longitude": longitude,
+                "image_id": None,
+                "image_path": None,
+            }
+
+        images = response.json()["data"]
+        log.debug("Successfully Retrieved Image Data: %s", images)
+        if len(images) == 0:
+            log.debug(
+                "No Images in Bounding Box: %s", self._bounds(latitude, longitude)
+            )
+            return {
+                "latitude": latitude,
+                "longitude": longitude,
+                "image_id": None,
+                "image_path": None,
+            }
+
+        image_id = images[0]["id"]
+        image_url = images[0]["thumb_original_url"]
+        image_path = self._download_image(image_url, image_id)
+
+        return {
+            "latitude": latitude,
+            "longitude": longitude,
+            "image_id": image_id,
+            "image_path": str(image_path),
+        }
+
+    def _bounds(self, latitude, longitude) -> str:
+        left = longitude - 10 / 111_111
+        bottom = latitude - 10 / 111_111
+        right = longitude + 10 / 111_111
+        top = latitude + 10 / 111_111
+        return f"{left},{bottom},{right},{top}"
+
+    def _download_image(self, image_url, image_id) -> Optional[Path]:
+        log.info("Downloading Image: %s", image_id)
+        try:
+            response = self.client.get(image_url, stream=True)
+            response.raise_for_status()
+        except RequestException as e:
+            log.error(e)
+            return None
+        image_content = response.content
+        log.debug("Successfully Retrieved Image: %s", image_id)
+        image_path = Path(self.basepath, f"{image_id}.jpeg")
+        log.debug("Writing Image To: %s", image_path)
+
+        if not image_path.is_file():
+            with open(image_path, "wb") as img:
+                img.write(image_content)
+            log.debug("Successfully Wrote Image: %s", image_path)
+
+        return image_path
+
+
+@app.command()
+def main(
+    client_token: Annotated[
+        str,
+        Argument(help="Mapillary Client Token from Developer Portal"),
+    ],
+    points_file: Annotated[
+        Path,
+        Argument(help=("Path to Input Points File")),
+    ],
+    image_path: Annotated[
+        Path,
+        Argument(help="Folder to Write Image Data"),
+    ] = Path(Path(__file__).parent.parent, "data/raw/mapillary"),
+    verbose: Annotated[bool, Option] = False,
+):
+    if verbose:
+        log.setLevel(logging.DEBUG)
+
+    if points_file.suffix == ".gpkg":
+        parser = GeoPandasParser(points_file)
+    else:
+        raise ValueError(f"Unsupported File Extension: {points_file.suffix}")
+
+    mapillary = Mapillary(client_token, image_path)
+    coordinates = parser.get_coordinates()
+
+    with Pool() as pool:
+        image_data = list(pool.map(mapillary.get_image_from_coordinates, coordinates))
+    log.debug(image_data)
```

Comment:
If we want to extend our scripts in the future by adding more data sources or image sources, having a separate GeoPandas parser will be crucial for parsing standardization and code reuse.

Similarly, if we ever have scripts that output different data file formats (JSON, XML, etc), we can determine what type of parser we want to use by checking the file extension. IIRC the ability to modularize and use different scripts based on what data/images we already have and what data/images we need is necessary functionality for this set of scripts.

Reply:
A few thoughts here:

1. geopandas itself is already that abstraction layer. It supports reading a variety of standard GIS data file formats like shapefile, GeoJSON, GeoPackage, etc. into an common class (`GeoDataFrame`) with a nice standard interface. geopandas is also a standard tool that other people who do Python GIS work already know, while a custom parser is a new thing that developers have to onboard onto. 
2. We control the data generation process here, and we should stick with a normal standard GIS data format. Per #16, we've decided to standardize on GeoPackage.  
3. Even if it were the case that we wanted to put another abstraction layer here (which I don't think makes sense), this feels like significant premature optimization. This is 30-40 lines of code to replace 3 lines of code for a use case that doesn't exist (we don't have any exotic XML data that we need to support loading).

Reply:
I can get behind only using GeoPackage as the only data file format, but there should still be a separate abstraction layer for common GeoPackage parsing functionality for the following reasons:

1. At a fundamental level, structuring things this way supports separation of concerns. A mapillary image downloader should not have to know about the content of a GeoPackage/GeoDataFrame or which column of the GeoDataFrame contains the coordinates to download images for. That logic is relegated to a common parser which all image downloaders can use.
2. More practically, this better supports modularity and code reuse. As stated earlier, I recall that a future goal of this was to be able to use different scripts to download images from different sources, get data from different sources, and process data in different ways. Having separate modules for each concern supports this, and helps maintainability so that if we want to change the structure of the points file in the future, we don't have to update the parsing logic in multiple places. In my mind, it's not a lot of work for a large amount of futureproofing, and it's not necessarily premature optimization if it's a feature that has been mentioned before.

I think of this parser not as a parser for GeoPackages, but as a parser for the GeoDataFrames that we export in the format/table structure that we have chosen.

Reply:
I'm going to make one more attempt here to argue my case. However, this isn't a hill I need to die on, so if you absolutely still disagree, then I'm willing to just agree to disagree and we go with your implementation. 

> there should still be a separate abstraction layer for common GeoPackage parsing functionality

geopanda's `GeoDataFrame` _is_ already such an abstraction layer and that's its entire purpose of existence. You can read all of the common GIS data file formats into a `GeoDataFrame` using a single function (`geopandas.read_file`). It provides a single interface no matter the on-disk data file format (the data will always be in a column named `"geometry"`, and will always be of Shapely's `Point` data type—this is a special fixed-name column in geopandas). 

Adding a customer parser class doesn't provide any improvement in realizing separation of concerns. The surface area of the interface is identical: you're replacing `gpd.read_file(...)` with GeoPandasParser(...)` and `gdf["geometry"]` with `parser.get_cordinates()`.

> A mapillary image downloader should not have to know about the content of a GeoPackage/GeoDataFrame or which column of the GeoDataFrame contains the coordinates to download images for.

You can use a `GeoDataFrame` and already not need to know anything about the content of the GeoPackage file on disk or the `GeoDataFrame`. The `"geometry"` column is a special column that all `GeoDataFrame`'s always have and will always contain the coordinates with a consistent type.

> As stated earlier, I recall that a future goal of this was to be able to use different scripts to download images from different sources, get data from different sources, and process data in different ways. 

Even if we were to get an arbitrary dataset of point coordinates (which I believe wouldn't happen anytime soon, given the relative newness of this project), it would trivially easy to parse it into a `GeoDataFrame` with geopandas using standard geopandas functionality. 

> Having separate modules for each concern supports this, and helps maintainability so that if we want to change the structure of the points file in the future, we don't have to update the parsing logic in multiple places. 

geopandas already gives us a lot of flexibility in changing the points file while needing zero other changes. We can change it to a shapefile, or a GeoJSON file, and the exact same `geopandas.read_file` call will produce an identical `GeoDataFrame`. And, it seems very unlikely to me that we'd ever want to support an exotic or bespoke data format that geopandas does not support. This is just a list of point coordinates, and there's not going to be any advantage to representing a list of point coordinates in a weird non-standard data format. 

> as a parser for the GeoDataFrames that we export in the format/table structure that we have chosen.

geopandas is already designed to be the standard data structure you use for all vector GIS data. It already serves the purpose that you are implementing a parser for.

Reply:
I think I understand your point, and I think I agree with most of it. Given that we also need to carry over data from the previous file, your approach makes more sense. Instead of framing this script as a Mapillary Client, I think framing it as an image downloader makes more sense, and we can add different image source clients as necessary. However, based on the data that needs to be added to the GeoDataFrame we will need to add arbitrary columns to the GeoDataFrame for the image processing results.